File: Java/pipeline/src/main/java/com/ibrahimatay/stage/RemoveDuplicatesStage.java
package com.ibrahimatay.stage;

import com.ibrahimatay.model.Message;
import com.ibrahimatay.model.Messages;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

public class RemoveDuplicatesStage implements Stage<Messages> {
    @Override
    public Messages execute(Messages input) {
        Set<Message> messages = new HashSet<>();
        messages.addAll(input.messages());

        return new Messages(new ArrayList<>(messages));
    }
}

File: Csharp/DesignPatterns.Pipeline/Stage/RemoveDuplicatesStage.cs
using DesignPatterns.Pipeline.Model;

namespace DesignPatterns.Pipeline.Stage;

public class RemoveDuplicatesStage:IStage<Messages>
{
    public Messages Execute(Messages input)
    {
        List<Message> messages = new List<Message>();
        foreach (var message in input.MessageList)
        {
            if (!messages.Contains(message))
            {
                messages.Add(message);
            }
        }
        
        return new Messages(messages);
    }
}